src/stream/src/executor/sink.rs

```diff
@@ -557,7 +607,7 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
         columns: Vec<ColumnCatalog>,
         mut sink_param: SinkParam,
         mut sink_writer_param: SinkWriterParam,
-        pk_matched: bool,
+        upsert_behavior: Option<UpsertBehavior>,
```

Comment:
In original code, when both `stream_key` and `user specified pk` is empty, `pk_mismatched` will be false and `pk_matched` will be true, while the current `Some(ub) && ub.should_compact_in_log_reader()` will be false, and the logic becomes different.

In our system we can assume that `stream_key` must not be empty. So we may add an assertion in the caller of this method that `stream_key` is non-empty, and then we can reason that the original logic is the same as current logic.